eureka2-core/src/main/java/com/netflix/eureka2/registry/NotifyingInstanceInfoHolder.java

```diff
@@ -255,4 +274,94 @@ public void shutdown() {
             super.shutdown();
         }
     }
+
+    /**
+     * Assume access to this is synchronized.
+     * All access to the dataStore first route through the sourceMap for the authoritative source to use
+     */
+    /* visible for testing */ static class DataStore {
+        protected final Map<String, Source> sourceMap = new HashMap<>();
+        protected final LinkedHashMap<Source, InstanceInfo> dataMap = new LinkedHashMap<>();  // for ordering
+
+        /**
+         * Matches sources on origin:name only.
+         * - If a matching source already exist in the sourceMap, remove from the dataStore first before add (as the
+         *   curr source may not equal due to a different source id).
+         * - Otherwise, just add to the dataStore
+         * - finally, add the source and sourceKey to the sourceMap
+         */
+        public void put(Source source, InstanceInfo instanceInfo) {
+            String sourceKey = sourceKey(source);
+            Source currIfExist = sourceMap.get(sourceKey);
+            if (currIfExist != null) {
+                dataMap.remove(currIfExist);
```

Comment:
Unfortunately this is not trivial unless we override Source's equals and hashcode to ignore the id, which come with its own set of issues. The whole ordering of the copies is just an optimisation anyway, so I see no strong need for it to be enforced.